crates/rust-client/src/store/note_record/input_note_record.rs

```
@@ -32,277 +32,306 @@ use crate::ClientError;
-        self.inclusion_proof = inclusion_proof;
+    /// Returns whether the note record contains a valid inclusion proof correlated with its
+    /// status
+    pub fn is_authenticated(&self) -> bool {
+        matches!(
+            self.state,
+            NoteState::Committed { .. }
+                | NoteState::ProcessingAuthenticated { .. }
+                | NoteState::NativeConsumed { .. }
+        )
     }
 
-    pub fn ignored(&self) -> bool {
-        self.ignored
+    // TRANSITIONS
+    // ================================================================================================
+
+    pub fn inclusion_proof_received(
+        &mut self,
+        inclusion_proof: NoteInclusionProof,
+        metadata: NoteMetadata,
+    ) -> Result<bool, NoteRecordError> {
+        match &self.state {
+            // Note had no inclusion proof
+            NoteState::Expected { .. } | NoteState::Unknown => {
+                self.state = NoteState::Unverified { inclusion_proof, metadata };
+                Ok(true)
+            },
+            // Note had an inclusion proof
+            NoteState::Unverified {
+                metadata: old_metadata,
+                inclusion_proof: old_inclusion_proof,
+            }
+            | NoteState::Committed {
+                metadata: old_metadata,
+                inclusion_proof: old_inclusion_proof,
+                ..
+            }
+            | NoteState::ProcessingAuthenticated {
+                metadata: old_metadata,
+                inclusion_proof: old_inclusion_proof,
+                ..
+            } => {
+                if old_inclusion_proof != &inclusion_proof || old_metadata != &metadata {
+                    return Err(NoteRecordError::StateTransitionError(
+                        "Inclusion proof or metadata do not match the expected values".to_string(),
+                    ));
+                }
+                Ok(false)
+            },
```

I can see how these statements can become pretty difficult to read - but not sure what the best solution is. One option is to introduce a concrete struct for reach variant - e.g., `ExpectedNoteState`, `UnverifiedNoteState` etc. -then, this would be just a dispatch to the underlying concrete structs. But the code may become quite verbose (though, maybe that's a good tradeoff).

---

Yes, with the number of states we have now these match cases can get pretty big. They will get a little larger as I'm planning to remove the `_` default case (I feel like it hides some transitions, we may want them to be explicit so that every transition is clear and when a new state is added it doesn't automatically fall in the default state).

I'll try the dispatch option and see how it looks.